app/grandchallenge/notifications/migrations/0008_migrate_forum_follows_and_notifications.py

```diff
@@ -0,0 +1,111 @@
+
+from grandchallenge.notifications.models import NotificationTypeChoices
+
+
+def migrate_forum_and_topic_follows_and_notifications(apps, schema_editor):
+    Notification = apps.get_model(  # noqa: N806
+        "notifications", "Notification"
+    )
+    Follow = apps.get_model("actstream", "Follow")  # noqa: N806
+    Forum = apps.get_model("discussion_forums", "Forum")  # noqa: N806
+    ForumTopic = apps.get_model(  # noqa: N806
+        "discussion_forums", "ForumTopic"
+    )
+    MachinaForum = apps.get_model("forum", "Forum")  # noqa: N806
+    MachinaTopic = apps.get_model("forum_conversation", "Topic")  # noqa: N806
+    ContentType = apps.get_model("contenttypes", "ContentType")  # noqa: N806
+
+    new_topic_ct = ContentType.objects.get_for_model(ForumTopic)
+    new_forum_ct = ContentType.objects.get_for_model(Forum)
+
+    def get_matching_forum(*, old_forum_id):
+        old_forum = MachinaForum.objects.get(pk=old_forum_id)
+        return old_forum.challenge.discussion_forum
+
+    def get_matching_topic(*, old_topic_id):
+        old_topic = MachinaTopic.objects.get(pk=old_topic_id)
+        new_forum = get_matching_forum(old_forum_id=old_topic.forum.pk)
+        return ForumTopic.objects.get(
+            forum=new_forum,
+            creator=old_topic.poster,
+            subject=old_topic.subject,
+        )
+
+    notifications_to_update = []
+    for notification in Notification.objects.filter(
+        type__in=[
+            NotificationTypeChoices.FORUM_POST,
+            NotificationTypeChoices.FORUM_POST_REPLY,
+        ]
+    ):
+        if notification.type == NotificationTypeChoices.FORUM_POST:
+            new_target = get_matching_forum(
+                old_forum_id=notification.target_object_id
+            )
+            new_action_object = get_matching_topic(
+                old_topic_id=notification.action_object_object_id
+            )
+            notification.target_content_type = new_forum_ct
+            notification.target_object_id = new_target.pk
+            notification.action_object_content_type = new_topic_ct
+            notification.action_object_object_id = new_action_object.pk
+        else:
+            new_target = get_matching_topic(
+                old_topic_id=notification.target_object_id
+            )
+            notification.target_content_type = new_topic_ct
+            notification.target_object_id = new_target.pk
+
+        notifications_to_update.append(notification)
+
+    Notification.objects.bulk_update(
+        notifications_to_update,
+        fields=[
+            "target_content_type",
+            "target_object_id",
+            "action_object_content_type",
+            "action_object_object_id",
+        ],
+        batch_size=1000,
+    )
+
+    ct_forum = ContentType.objects.get_for_model(MachinaForum)
+    ct_forumtopic = ContentType.objects.get_for_model(MachinaTopic)
+
+    follows_to_update = []
+    for follow in Follow.objects.filter(
```

Comment:
Follow is missing from the Django admin so I cannot get insights into how much data we have there, but for the object permissions it looks like there are about 1M.

Reply:
There is a `Follows` admin in the Django admin. There are 1148776 objects, but I can't filter by content type easily, so not sure how many of those would need to be updated. In any case, a lot though.

Reply:
So this also needs batching here, and not just while updating below. Follows do need to be updated, no option to just delete those.

Reply:
I have added the batching here. I also checked and ECS tasks do not have a time limit, so this should work.